rfcs/proposed/task_group_dynamic_dependencies/extended_semantics.md

```diff
@@ -0,0 +1,1453 @@
+
+<img src="assets/parser_tasks1.png" width=400>
+
+`f4p` and `f3p` are at least in `submitted` state when the edge with `f5f` is created.
+
+Let's assume that the task `f3p` is executed next. It creates the finalize task `f3f` and two tasks for includes - `f2p` and `f1p`:
+
+<img src="assets/parser_tasks2.png" width=400>
+
+After creating the edges between includes and `f3f` are added, `f3p` transfers it's successors to `f3f` and destroys:
+
+<img src="assets/parser_tasks3.png" width=400>
+
+Assume that `f4p` is taken for execution next. It creates finalize task `f4f` and should create edges for with *File 3* and *File 2*.
+
+Task for *File 2* was already created when parsing *File 3*, so the tracker to `f2p` would be found in the `fileMap`. 
+
+The tracker for *File 3* in the `fileMap` was initially set to track `f3p`, but it have already transferred it's successors to `f3f`, so
+new edges needed to be added to `f3f`, but using the tracker to `f3p`:
+
+<img src="assets/parser_tasks4.png" width=400>
+
+Similar to `f3p`, `f4p` transfers it's successor `f5f` to finalize task `f4f`:
+
+<img src="assets/parser_tasks5.png" width=400>
+
+When `f2p` is executed, it would create a finalize task `f2f`, connects it to `f1p` and transfers it's successors to `f2f`:
+
+<img src="assets/parser_tasks6.png" width=400>
+
+The last task to be executed is `f1p` that just replaces itself with the finalize task since there are no dependencies:
+
+<img src="assets/parser_tasks7.png" width=400>
+
+After all of the dependencies are made, *File 1* is finalized, than the *File 2*, *File 3*, *File 4* and *File 5*
+
+## API proposal summary
+
+```cpp
+namespace oneapi {
+namespace tbb {
+
+// Existing API
+class task_handle;
+
+// New APIs
+class task_tracker {
+public:
+    task_tracker();
+
+    task_tracker(const task_tracker& other);
+    task_tracker(task_tracker&& other);
+
+    task_tracker(const task_handle& handle);
+
+    ~task_tracker();
+
+    task_tracker& operator=(const task_tracker& other);
+    task_tracker& operator=(task_tracker&& other);
+
+    task_tracker& operator=(const task_handle& handle);
+
+    explicit operator bool() const noexcept;
+}; // class task_tracker
+
+bool operator==(const task_tracker& t, std::nullptr_t) noexcept;
+bool operator==(std::nullptr_t, const task_tracker& t) noexcept;
+
+bool operator!=(const task_tracker& t, std::nullptr_t) noexcept;
+bool operator!=(std::nullptr_t, const task_tracker& t) noexcept;
+
+bool operator==(const task_tracker& t, const task_tracker& rhs) noexcept;
+bool operator!=(const task_tracker& t, const task_tracker& rhs) noexcept;
+
+class task_group {
+    static void make_edge(task_handle& pred, task_handle& succ);
+    static void make_edge(task_tracker& pred, task_handle& succ);
+
+    struct current_task {
+        static void transfer_successors_to(task_handle& h);
+    };
+}; // class task_group
```

Comment:
I think we should consider introducing the return of a `[[maybe_unused]]` `task_handle`/`task_tracker` by the `task_group::run(Func&& f)`?